File: backend/app/schemas/progress.py
from pydantic import BaseModel
from typing import Optional, Any, Dict, List
from datetime import datetime
from app.models.module import ModuleStatus, ChallengeStatus


class ProgressResponse(BaseModel):
    user_id: int
    module_id: int
    status: ModuleStatus
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ChallengeAttemptResponse(BaseModel):
    id: int
    user_id: int
    module_id: int
    challenge_id: int
    status: ChallengeStatus
    attempt_count: int
    validation_output: Optional[Dict[str, Any]] = None
    passed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ValidationRequest(BaseModel):
    module_id: int
    challenge_id: int
    payload: Optional[Dict[str, Any]] = None


class ValidationResponse(BaseModel):
    passed: bool
    message: str
    details: Optional[Dict[str, Any]] = None


class LabSessionResponse(BaseModel):
    id: int
    module_id: int
    container_id: Optional[str] = None
    container_name: Optional[str] = None
    container_port: Optional[int] = None
    container_host: Optional[str] = None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class QuizSubmission(BaseModel):
    module_id: int
    answers: Dict[int, Any]  # question_index -> answer


class QuizResultResponse(BaseModel):
    score: int
    max_score: int
    passed: bool
    correct_answers: Dict[int, Any]


class DashboardResponse(BaseModel):
    user_id: int
    username: str
    total_modules: int = 10
    completed_modules: int
    in_progress_modules: int
    overall_percentage: float
    module_progress: List[ProgressResponse]
    recent_activity: List[ChallengeAttemptResponse]

File: frontend/src/store/progress.ts
import { create } from "zustand";
import { progressApi, labsApi } from "@/lib/api";

export type ModuleStatus = "locked" | "available" | "in_progress" | "completed";
export type ChallengeStatus = "not_started" | "in_progress" | "passed" | "failed";

export interface ModuleProgress {
  user_id: number;
  module_id: number;
  status: ModuleStatus;
  started_at: string | null;
  completed_at: string | null;
}

export interface ChallengeAttempt {
  id: number;
  module_id: number;
  challenge_id: number;
  status: ChallengeStatus;
  attempt_count: number;
  validation_output: Record<string, unknown> | null;
  passed_at: string | null;
}

export interface LabSession {
  id: number;
  module_id: number;
  container_id: string | null;
  container_name: string | null;
  container_port: number | null;
  container_host: string | null;
  status: string;
  created_at: string;
}

interface ProgressState {
  modules: ModuleProgress[];
  challenges: Record<number, ChallengeAttempt[]>;
  activeSessions: Record<number, LabSession>;
  dashboardData: Record<string, unknown> | null;
  isLoading: boolean;

  fetchDashboard: () => Promise<void>;
  fetchModules: () => Promise<void>;
  fetchChallenges: (moduleId: number) => Promise<void>;
  startLab: (moduleId: number) => Promise<LabSession>;
  stopLab: (moduleId: number) => Promise<void>;
  validateChallenge: (
    moduleId: number,
    challengeId: number,
    payload?: Record<string, unknown>
  ) => Promise<{ passed: boolean; message: string; details: Record<string, unknown> }>;
}

export const useProgressStore = create<ProgressState>((set, get) => ({
  modules: [],
  challenges: {},
  activeSessions: {},
  dashboardData: null,
  isLoading: false,

  fetchDashboard: async () => {
    const res = await progressApi.dashboard();
    set({ dashboardData: res.data });
  },

  fetchModules: async () => {
    set({ isLoading: true });
    const res = await progressApi.modules();
    set({ modules: res.data, isLoading: false });
  },

  fetchChallenges: async (moduleId) => {
    const res = await progressApi.challenges(moduleId);
    set((s) => ({ challenges: { ...s.challenges, [moduleId]: res.data } }));
  },

  startLab: async (moduleId) => {
    const res = await labsApi.start(moduleId);
    const session: LabSession = res.data;
    set((s) => ({ activeSessions: { ...s.activeSessions, [moduleId]: session } }));
    return session;
  },

  stopLab: async (moduleId) => {
    await labsApi.stop(moduleId);
    set((s) => {
      const sessions = { ...s.activeSessions };
      delete sessions[moduleId];
      return { activeSessions: sessions };
    });
  },

  validateChallenge: async (moduleId, challengeId, payload) => {
    const res = await labsApi.validate(moduleId, challengeId, payload);
    // Re-fetch modules to get updated lock state
    await get().fetchModules();
    await get().fetchChallenges(moduleId);
    return res.data;
  },
}));
